Userland/Libraries/LibWeb/WebDriver/ExecuteScript.cpp

```diff
@@ -0,0 +1,363 @@
+#include <LibJS/Runtime/Promise.h>
+#include <LibJS/Runtime/PromiseConstructor.h>
+#include <LibWeb/DOM/Document.h>
+#include <LibWeb/DOM/HTMLCollection.h>
+#include <LibWeb/DOM/NodeList.h>
+#include <LibWeb/FileAPI/FileList.h>
+#include <LibWeb/HTML/BrowsingContext.h>
+#include <LibWeb/HTML/HTMLOptionsCollection.h>
+#include <LibWeb/HTML/Scripting/Environments.h>
+#include <LibWeb/HTML/Window.h>
+#include <LibWeb/Page/Page.h>
+#include <LibWeb/WebDriver/ExecuteScript.h>
+
+namespace Web::WebDriver {
+
+static ErrorOr<JsonValue, ExecuteScriptResultType> internal_json_clone_algorithm(JS::Realm&, JS::Value, HashTable<JS::Object*>& seen);
+static ErrorOr<JsonValue, ExecuteScriptResultType> clone_an_object(JS::Realm&, JS::Object&, HashTable<JS::Object*>& seen, auto const& clone_algorithm);
+
+// https://w3c.github.io/webdriver/#dfn-collection
+static bool is_collection(JS::Object const& value)
+{
+    // A collection is an Object that implements the Iterable interface, and whose:
+    return (
+        // - initial value of the toString own property is "Arguments"
+        value.has_parameter_map()
+        // - instance of Array
+        || is<JS::Array>(value)
+        // - instance of FileList
+        || is<FileAPI::FileList>(value)
+        // - instance of HTMLAllCollection
+        || false // FIXME
+        // - instance of HTMLCollection
+        || is<DOM::HTMLCollection>(value)
+        // - instance of HTMLFormControlsCollection
+        || false // FIXME
+        // - instance of HTMLOptionsCollection
+        || is<HTML::HTMLOptionsCollection>(value)
+        // - instance of NodeList
+        || is<DOM::NodeList>(value));
+}
+
+// https://w3c.github.io/webdriver/#dfn-json-clone
+static ErrorOr<JsonValue, ExecuteScriptResultType> json_clone(JS::Realm& realm, JS::Value value)
+{
+    // To perform a JSON clone return the result of calling the internal JSON clone algorithm with arguments value and an empty List.
+    auto seen = HashTable<JS::Object*> {};
+    return internal_json_clone_algorithm(realm, value, seen);
+}
+
+// https://w3c.github.io/webdriver/#dfn-internal-json-clone-algorithm
+static ErrorOr<JsonValue, ExecuteScriptResultType> internal_json_clone_algorithm(JS::Realm& realm, JS::Value value, HashTable<JS::Object*>& seen)
+{
+    auto& vm = realm.vm();
+
+    // When required to run the internal JSON clone algorithm with arguments value and seen, a remote end must return the value of the first matching statement, matching on value:
+    // -> undefined
+    // -> null
+    if (value.is_nullish()) {
+        // Success with data null.
+        return JsonValue {};
+    }
+
+    // -> type Boolean
+    // -> type Number
+    // -> type String
+    if (value.is_boolean() || value.is_number() || value.is_string()) {
+        // Success with data value.
+        auto json_string = MUST(JS::JSONObject::stringify_impl(vm, value, JS::js_undefined(), JS::js_undefined()));
+        return MUST(JsonValue::from_string(json_string));
```

Comment:
`Value` -> `String` -> `JsonValue` seems a bit excessive? At least for bools and numbers

Reply:
Fair enough, it's `Value` -> `JsonValue` now :)